pkg/flowaggregator/flowaggregator.go

```diff
@@ -325,46 +325,57 @@ func (fa *flowAggregator) initExportingProcess() error {
 }
 
 func (fa *flowAggregator) Run(stopCh <-chan struct{}) {
-	exportTicker := time.NewTicker(fa.exportInterval)
-	defer exportTicker.Stop()
 	go fa.collectingProcess.Start()
 	defer fa.collectingProcess.Stop()
 	go fa.aggregationProcess.Start()
 	defer fa.aggregationProcess.Stop()
+	go fa.flowRecordExpiryCheck(stopCh)
+
+	<-stopCh
+}
+
+func (fa *flowAggregator) flowRecordExpiryCheck(stopCh <-chan struct{}) {
+	expireTicker := time.NewTicker(fa.activeFlowRecordTimeout)
+
 	for {
 		select {
 		case <-stopCh:
 			if fa.exportingProcess != nil {
 				fa.exportingProcess.CloseConnToCollector()
 			}
+			expireTicker.Stop()
 			return
-		case <-exportTicker.C:
+		case <-expireTicker.C:
 			if fa.exportingProcess == nil {
 				err := fa.initExportingProcess()
 				if err != nil {
-					klog.Errorf("Error when initializing exporting process: %v, will retry in %s", err, fa.exportInterval)
-					// Initializing exporting process fails, will retry in next exportInterval
+					klog.Errorf("Error when initializing exporting process: %v, will retry in %s", err, fa.activeFlowRecordTimeout)
+					// Initializing exporting process fails, will retry in next cycle.
+					expireTicker.Stop()
+					expireTicker = time.NewTicker(fa.activeFlowRecordTimeout)
 					continue
 				}
 			}
-			err := fa.aggregationProcess.ForAllRecordsDo(fa.sendFlowKeyRecord)
-			if err != nil {
-				klog.Errorf("Error when sending flow records: %v", err)
+			// Pop the flow record item from expire priority queue in the Aggregation
+			// Process and send the flow records.
+			if err := fa.aggregationProcess.ForAllExpiredFlowRecordsDo(fa.sendFlowKeyRecord); err != nil {
+				klog.Errorf("Error when sending expired flow records: %v", err)
 				// If there is an error when sending flow records because of intermittent connectivity, we reset the connection
 				// to IPFIX collector and retry in the next export cycle to reinitialize the connection and send flow records.
 				fa.exportingProcess.CloseConnToCollector()
 				fa.exportingProcess = nil
+				expireTicker.Stop()
+				expireTicker = time.NewTicker(fa.activeFlowRecordTimeout)
```

Comment:
why `fa.activeFlowRecordTimeout`?

it seems that we could just remove the `continue` statement and let the code below handle resetting the timer?

Reply:
There can be error when sending flow records to the external collector and it may be down for some time, so do not want to retry immediately by picking the min expiry value. I wanted to reset it.

Reply:
sounds good